library/Stratosphere/Values.hs

```diff
@@ -22,38 +21,27 @@ import Data.String (IsString (..))
 import Data.Text (Text)
 import GHC.Exts (IsList(..))
 
--- GADTs are cool, but I couldn't get this to work with FromJSON. Now that we
--- don't have FromJSON though, we can revisit this.
-
--- data Val a where
---   Literal :: a -> Val a
---   Ref :: Text -> Val a
---   If :: Text -> Val a -> Val a -> Val a
---   And :: Val Bool -> Val Bool -> Val Bool
---   Equals :: (Show a, ToJSON a) => Val a -> Val a -> Val Bool
---   Or :: Val Bool -> Val Bool -> Val Bool
-
 -- | This type is a wrapper around any values in a template. A value can be a
 -- 'Literal', a 'Ref', or an intrinsic function. See:
 -- http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference.html
-data Val a
- = Literal a
- | Ref Text
- | If Text (Val a) (Val a)
- | And (Val Bool) (Val Bool)
- | Equals (Val a) (Val a)
- | Or (Val Bool) (Val Bool)
- | GetAtt Text Text
- | Base64 (Val Text)
- | Join Text (ValList Text)
- | Select Integer (ValList a)
- | FindInMap (Val a) (Val a) (Val a) -- ^ Map name, top level key, and second level key
- | ImportValue Text -- ^ The account-and-region-unique exported name of the value to import
- | Sub Text (Maybe (HashMap Text (Val Text))) -- ^ Substitution string and optional map of values
+data Val a where
+  Literal :: a -> Val a
+  Ref :: Text -> Val a
+  If :: Text -> Val a -> Val a -> Val a
+  And :: Val Bool' -> Val Bool' -> Val Bool'
+  Equals :: (Show a, ToJSON a) => Val a -> Val a -> Val Bool'
+  Or :: Val Bool' -> Val Bool' -> Val Bool'
+  GetAtt :: Text -> Text -> Val a
+  Base64 :: Val Text -> Val Text
+  Join :: Text -> ValList Text -> Val Text
+  Select :: Integer -> ValList a -> Val a
+  FindInMap :: Val a -> Val a -> Val a -> Val a -- ^ Map name, top level key, and second level key
+  ImportValue :: Text -> Val a -- ^ The account-and-region-unique exported name of the value to import
+  Sub :: Text -> Maybe (HashMap Text (Val Text)) -> Val Text -- ^ Substitution string and optional map of values
 
 deriving instance (Show a) => Show (Val a)
-deriving instance (Eq a) => Eq (Val a)
-deriving instance Functor Val
+
+-- N.B. The 'Equals' function makes it so we can't derive 'Eq'
```

Comment:
Does adding an `Eq` constraint to the existential (where you have `(Show a, ToJSON a)`) not work?

Reply:
Nope :( I haven't worked on this too hard either. I just did the obvious transformation and I was going to let this sit for a bit while we though about it :smile:

Reply:
I don't know that it's possible to write a `Functor` instance for this since the type param can't be changed for some constructors, but it is possible to write `Eq`:

```haskell
data Val a where
  Literal :: a -> Val a
  ...
  Equals :: (Typeable x, Eq x, Show x, ToJSON x) => Val x -> Val x -> Val Bool'
  ...

instance Eq a => Eq (Val a) where
  Literal a == Literal b = a == b
  ...
  Equals a b == Equals c d = eqEquals a b c d
  _ == _ = False

eqEquals :: (Typeable a, Typeable b, Eq a, Eq b) => a -> a -> b -> b -> Bool
eqEquals a b c d = fromMaybe False $ do
  c' <- cast c
  d' <- cast d
  pure $ a == c' && b == d'
```